### PR TITLE
CRM-19330 Add Contacts to Group appears and Action do nothing for non-privileged users.

### PR DESCRIPTION
**Pull request by spalmstr**
Add Contacts to Group appears and Action do nothing for non-privileged
users.

---
- [CRM-19330: Add Contacts to Group appears and Action do nothing for non-privileged users under Joomla](https://issues.civicrm.org/jira/browse/CRM-19330)

### CODE REVIEW

File: CRM/Report/Form.php [hunk 1]
```diff
@@ -4779,6 +4779,16 @@ protected function setOutputMode() {
...
+    if ($this->_outputMode !== NULL) {
+      // We didn't get it using output, so we now try with task.
+      $this->_outputMode = str_replace('report_instance.', '', CRM_Utils_Request::retrieve(
+          'task',
+          'String',
+          CRM_Core_DAO::$_nullObject,
+          FALSE,
+          CRM_Utils_Array::value('task', $this->_params)
+        ));
+    }
```
Comment (jitendrapurohit) [inline CRM/Report/Form.php]: The comment and the if condition looks contradictory. Also, I think the `_outputMode` should be set above itself when we get `task` in `$this->_params` if we didn't get it using the `output` request. 

It is set correctly for me due to which I was able to perform the remaining task(print, export, pdf) without having `administer report` permission. Is there something I'm missing ?

File: CRM/Report/Form.php [hunk 2]
```diff
@@ -1498,9 +1498,9 @@ public function buildInstanceAndButtons() {
 
     $this->assign('instanceForm', $this->_instanceForm);
 
-    // CRM-16274 Determine if user has 'edit all contacts' or equivalent
-    $permission = CRM_Core_Permission::getPermission();
-    if ($permission == CRM_Core_Permission::EDIT &&
+    // CRM-19330 - check that the user has Edit All Groups permissions.
+    $permission = CRM_Core_Permission::check(CRM_Core_Permission::EDIT_GROUPS);
+    if ($permission  &&
```
Comment (spalmstr) [inline CRM/Report/Form.php]: I may have misunderstood the meaning of Edit All Groups.

File: CRM/Report/Form.php (hunk 1 above)
Comment (spalmstr) [inline CRM/Report/Form.php]: I have just tried with our test system running 4.7.13 with a non-privileged logon, and the various tasks did nothing, but just returned you to the report.  Saving the report left the user with a blank report.  Switching to the custom code results in.
The Add Contact to Group option disappearing and the various actions working.  The Save action are no longer visible.

File: CRM/Report/Form.php (hunk 2 above)
Comment (jitendrapurohit) [inline CRM/Report/Form.php]: Okay, so can you please update the PR removing this part ?

File: CRM/Report/Form.php (hunk 1 above)
Comment (jitendrapurohit) [inline CRM/Report/Form.php]: As `Save` and `Save a Copy` task has been removed from this PR, I was concerned about the remaining three actions(print, export, pdf) which seemed to work fine for me without this part. Can you please check and update? 
`Add Contact to Group` option disappearing is also an expected behaviour as per [above discussion](https://github.com/civicrm/civicrm-core/pull/9202/files#r90634203) .

Comment (spalmstr) [inline CRM/Report/Form.php]: Thanks for the feedback.  I am tied up until after Christmas, so any changes will have to wait till the New Year, I'm afraid.  I daren't make a quick change that does not work properly.

**Comment by spalmstr** [comment]
Merge build fails because a report file is missing. Otherwise it looks as though the PR has been updated.
I have, I hope, tweaked form.php so it does the right permissions check and ReportInstance.php so it suppresses Save and Save As unless the user has the right to edit reports.

**Comment by jitendrapurohit** [comment]
looks like this PR contains some false commits. Can you rebase and push again ? 

    git pull --rebase upstream master
    git push origin <your_branch>

**Comment by spalmstr** [comment]
Here is what happened when I tried the pull rebase and push.

Welcome to Git (version 1.9.5-preview20141217)


Run 'git help git' to display the help index.
Run 'git help <command>' to display help for specific commands.

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)
$ git pull --rebase upstream master
remote: Counting objects: 130, done.
remote: Compressing objects: 100% (49/49), done.
remote: Total 130 (delta 84), reused 64 (delta 64), pack-reused 16R
Receiving objects: 100% (130/130), 75.45 KiB | 0 bytes/s, done.
Resolving deltas: 100% (85/85), completed with 39 local objects.
From https://github.com/civicrm/civicrm-core
 * branch            master     -> FETCH_HEAD
   df7b228..e2b0438  master     -> upstream/master
First, rewinding head to replay your work on top of it...
Applying: CRM-19330
Applying: CRM-19330
Applying: CRM-19330
Applying: CRM-19330
Applying: CRM-19330
Applying: CRM-19330
Using index info to reconstruct a base tree...
M       CRM/Report/Form.php
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)
$ git push origin CRM-19330
To https://spalmstr@github.com/spalmstr/civicrm-core.git
 ! [rejected]        CRM-19330 -> CRM-19330 (non-fast-forward)
error: failed to push some refs to 'https://spalmstr@github.com/spalmstr/civicrm-core.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)
$ git pull origin CRM-19330
From https://github.com/spalmstr/civicrm-core
 * branch            CRM-19330  -> FETCH_HEAD
Merge made by the 'recursive' strategy.

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)
$ git push origin CRM-19330
Counting objects: 474, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (104/104), done.
Writing objects: 100% (175/175), 51.40 KiB | 0 bytes/s, done.
Total 175 (delta 136), reused 95 (delta 68)
remote: Resolving deltas: 100% (136/136), completed with 60 local objects.
To https://spalmstr@github.com/spalmstr/civicrm-core.git
   d04adc9..8431001  CRM-19330 -> CRM-19330

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)
$

I hope this resolved the issue.

**Comment by spalmstr** [comment]
Hi,

 

Many thanks – I still think I am feeling my way around the complexities of Git.  Here is what happened when I tried pushing again.

 

Welcome to Git (version 1.9.5-preview20141217)

 

 

Run 'git help git' to display the help index.

Run 'git help <command>' to display help for specific commands.

 

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)

$ git pull --rebase upstream master

remote: Counting objects: 130, done.

remote: Compressing objects: 100% (49/49), done.

remote: Total 130 (delta 84), reused 64 (delta 64), pack-reused 16R

Receiving objects: 100% (130/130), 75.45 KiB | 0 bytes/s, done.

Resolving deltas: 100% (85/85), completed with 39 local objects.

From https://github.com/civicrm/civicrm-core

* branch            master     -> FETCH_HEAD

   df7b228..e2b0438  master     -> upstream/master

First, rewinding head to replay your work on top of it...

Applying: CRM-19330

Applying: CRM-19330

Applying: CRM-19330

Applying: CRM-19330

Applying: CRM-19330

Applying: CRM-19330

Using index info to reconstruct a base tree...

M       CRM/Report/Form.php

Falling back to patching base and 3-way merge...

No changes -- Patch already applied.

 

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)

$ git push origin CRM-19330

To https://spalmstr@github.com/spalmstr/civicrm-core.git

! [rejected]        CRM-19330 -> CRM-19330 (non-fast-forward)

error: failed to push some refs to 'https://spalmstr@github.com/spalmstr/civicrm-core.git'

hint: Updates were rejected because the tip of your current branch is behind

hint: its remote counterpart. Integrate the remote changes (e.g.

hint: 'git pull ...') before pushing again.

hint: See the 'Note about fast-forwards' in 'git push --help' for details.

 

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)

$ git pull origin CRM-19330

From https://github.com/spalmstr/civicrm-core

* branch            CRM-19330  -> FETCH_HEAD

Merge made by the 'recursive' strategy.

 

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)

$ git push origin CRM-19330

Counting objects: 474, done.

Delta compression using up to 4 threads.

Compressing objects: 100% (104/104), done.

Writing objects: 100% (175/175), 51.40 KiB | 0 bytes/s, done.

Total 175 (delta 136), reused 95 (delta 68)

remote: Resolving deltas: 100% (136/136), completed with 60 local objects.

To https://spalmstr@github.com/spalmstr/civicrm-core.git

   d04adc9..8431001  CRM-19330 -> CRM-19330

 

Stephen@JUPITER /d/CiviCRM_Custom.git/joomla/administrator/components/com_civicrm/civicrm/CRM/Report (CRM-19330)

$

 

I hope this has done the necessary.

 

Kind regards and Happy New Year.

 

Stephen Palmstrom

**Comment by jitendrapurohit** [comment]
Nope, seems like this one got messed up. Can you please create a fresh PR with your changes and close this ?

Thanks:-)

**Comment by spalmstr** [comment]
Closing and recreating, as it look as though this one got messed up.

**Pull request closed without merging**